### PR TITLE
[JENKINS-60902] Upgrade to SSHD 2.5.1 (Runnable version)

### PR DESCRIPTION
**Pull request by kuisathaverat**
see [JENKINS-60902](https://issues.jenkins-ci.org/browse/JENKINS-60902)

It bumps the [Apache Mina sshd](https://github.com/apache/mina-sshd) library to 2.5.1

it replaces https://github.com/jenkinsci/sshd-plugin/pull/45
it depends on [JENKINS-55582](https://issues.jenkins-ci.org/browse/JENKINS-55582)

This PR is the same change in https://github.com/jenkinsci/sshd-plugin/pull/37 but implement AsynchronousCommand as runnable due to some changes in the Apache Mina library, this requires changes on other plugins because of signature changes. The change is to rename the method `run` to `runCommand`.

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Review by jglick — COMMENTED** [review]
OK I think, so long as the `runCommand` change works out in downstream PRs.

**Comment by kuisathaverat** [comment]
The only read dependency of the plugin is in the git-server-plugin, in ssh-agent-plugin and gerrit-trigger-plugin  the dependency is with the Apache Minda SSHD library on tests.
I have tested git-server-plugin manually and the changes work. For ssh-agent-plugin I have made a PR that replaces the dependency of sshd-plugin with the Apache Mina SSHD library 2.5.1 when the Jenkins core dependency is set up to 2.282+.
For the gerrit-trigger-plugin is more complicated to make the same change, I have made a PR to support incremental (merged), another to use BOM (some test fails), and another one to replace the sshd-plugin dependency with the Apache Mina SSHD library 1.7.0 (same problems that the BOM pr). Due to this plugin uses the sshd-plugin only for import the Apache Mina SSHD library for testing, it does not have issues until it bumps the Jenkins core version up to 2.282+, and when this bump would make the plugins will have the issue I have had to use BOM and update the dependencies, this is a lot of work that I cannot make because I do not know the plugin well. So I'll leave that work for the maintainers when they bump the version up to 2.282+ they have halfway done on my PRs.

* https://github.com/jenkinsci/git-server-plugin/pull/24
* https://github.com/jenkinsci/ssh-agent-plugin/pull/54
* https://github.com/jenkinsci/gerrit-trigger-plugin/pull/441
* https://github.com/jenkinsci/gerrit-trigger-plugin/pull/442
* https://github.com/jenkinsci/gerrit-trigger-plugin/pull/443/

If there is no objection I would like to merge this PR and release a new minor version 3.1.0, I suggest a minor version due to it contains dependency changes, it is not a simple patch.

**Comment by timja** [comment]
Fine with me

**Comment by jglick** [comment]
If you know what you are doing, go ahead, and let me know if/when you need help.

**Comment by kuisathaverat** [comment]
@jglick we have to coordinate the merge and release of this PR and the merge of https://github.com/jenkinsci/git-server-plugin/pull/24 and the release of git-server-plugin to avoid break that plugin.

**Comment by jglick** [comment]
> we have to coordinate

Do you have permissions to merge & release on both? If not, have you secured a promise from someone who does to do the coördinated release?

**Review by jtnord — APPROVED** [review]
seems ok.

I notices that the weeklies of Jenkins are still using `2.6.0` and not `2.5.1` (as the LTS version) or `2.7.0` (the upstream fixed version)

File: src/main/java/org/jenkinsci/main/modules/sshd/AsynchronousCommand.java [hunk 1]
```diff
@@ -95,46 +96,54 @@ public Environment getEnvironment() {
...
+            int i;
+
+            // run the command in the context of the authenticated user
+            SecurityContext old = SecurityContextHolder.getContext();
+            User user = getCurrentUser();
+            if (user!=null)
+                ACL.impersonate(user.impersonate());
+
+            try {
+                i = AsynchronousCommand.this.runCommand();
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/main/modules/sshd/AsynchronousCommand.java:123]: is it not possible to change this to be backwards compatible - have you run usage-in-plugins against CloudBees and Jenkins update centers?  (I know there is at least one internal use of this).

Comment (timja) [reply src/main/java/org/jenkinsci/main/modules/sshd/AsynchronousCommand.java:123]: previous discussion: https://github.com/jenkinsci/sshd-plugin/pull/37#discussion_r624502145

Comment (jtnord) [reply src/main/java/org/jenkinsci/main/modules/sshd/AsynchronousCommand.java:123]: ok.  that list is not complete :)

**Comment by jglick** [comment]
Anything happening here?

**Comment by kuisathaverat** [comment]
nothing, I haven`t time to merge and release in the last two weeks, I will merge tomorrow and coordinate with the owner of the git-server plugin to release at the same time.

**Comment by kuisathaverat** [comment]
shit, I did not squash the PR 🤦

**Comment by jglick** [comment]
No harm done.

**Comment by jglick** [comment]
Neither this nor the downstream `git-server` seems to be listed on the UC for 2.289.1. You might want to check the UC generation logs to see if anything is wrong.

**Comment by timja** [comment]
https://gist.github.com/timja/0eaa47b442f30e71b7dc668266b68deb

**Comment by kuisathaverat** [comment]
I doubled check that it is in the artifactory https://repo.jenkins-ci.org/artifactory/releases/org/jenkins-ci/modules/sshd/3.0.4/

**Pull request merged**